### ISSUE TITLE
variable cards lost their names on client side

### ISSUE THREAD

**Issue by druidbruce** [CyclopsMC/IntegratedDynamics#1038]
#### Issue type:

- :bug: Bug 

____

#### Short description:

variable cards lost their names on client side

#### Steps to reproduce the problem:

1. not really sure how to reproduce this problem, but somehow the client side does not show the names
2. the names still actually exists since when you try to rename them in an anvil, it does not let you rename to the original name
3. if you put it in  operator `ToString :: Named -> String`, it does properly show its name
4. restarting client/server doesn't seem to fix the problem, it just somehow happened..

#### Expected behaviour:

on client side it shows the name
____

#### Versions:

- This mod: 
clinet same as server
CommonCapabilities-1.16.5-2.6.0
CyclopsCore-1.16.5-1.11.5
IntegratedCrafting-1.16.5-1.0.13
IntegratedDynamics-1.16.5-1.8.0
IntegratedTerminals-1.16.5-1.2.2
IntegratedTunnels-1.16.5-1.7.1
- Minecraft: 
1.16.5
- Forge: 
36.1.2

**Comment by rubensworks** [comment]
Thanks for reporting!

**Comment by druidbruce** [comment]
not sure if this matters, but there are currently 1697 variable cards on the server and most of them are named.

**Comment by druidbruce** [comment]
seems to be similar to #712 and #703, and seemed to be the same type of problem, where the labelled names are actually there but just not sent to client. 
Also, i just noticed another issue, where if you name a card in an anvil, you can no longer rename it in the labeller on the client side? what happen is this, the labeller changed name shows up in a card thats reference the changed card, but the official name(the one you get from `ToString :: Named -> String`) of the card does not change. not sure if this is intended

maybe the issue got reintroduced when updating to 1.8.0?

**Comment by rubensworks** [comment]
@druidbruce When did this start happening? After an update? After labeling a variable? After restarting the game? After relogging?

Does the problem also occur in new worlds?

**Comment by druidbruce** [comment]
> When did this start happening?

it started happening after the server updated to 1.8.0, but the thing is it didn't happen immediately, only after a few restarts(server).

This is the full story of what happened: server update to 1.8.0, restarted 2 times, I log off, (server restarted more times when i wasn't online), then when i logged back on the server the next day, all variable card names don't show directly on the client side. Tested this with another player as well.

> Does the problem also occur in new worlds?

not sure if this is what you mean, but doesn't seem to happen in single player

**Comment by druidbruce** [comment]
new observation:
If I keep client open as server restart, newly named cards(via labeller) do not forget their names(client side).
If I restart client after server restart, all named cards(via labeller) forgets their names(client side)

This explains why this happens
>This is the full story of what happened: server update to 1.8.0, restarted 2 times, I log off, (server restarted more times when i wasn't online), then when i logged back on the server the next day, all variable card names don't show directly on the client side. Tested this with another player as well.

as I did not close the client when the server restarted and update to 1.8.0(because client is already on 1.8.0), but once I log off and close client, the names are no longer kept.

**Comment by rubensworks** [comment]
@druidbruce Do you see any errors/warnings in the client/server logs? (feel free to send me the full logs)
At player login, the server should send all labels to the client, so then is the time at which something may be printed in the logs if anything goes wrong.

The 1.8.0 update changed nothing regarding labels, so I suspect it's either a corrupt label, or related to the (large) number of labels you have.

In any case, without logs, there's not much I can do about this at the moment I'm afraid :-(

**Comment by druidbruce** [comment]
@rubensworks 
this is the log where i simply launched the client and logged on the server
[latest.log](https://github.com/CyclopsMC/IntegratedDynamics/files/6454637/latest.log)

**Comment by rubensworks** [comment]
@druidbruce Thanks, and what about the server logs? There's nothing relevant in the client logs unfortunately.

**Comment by gjerm** [comment]
I'm having a similar issue, and the server log displays this on a player join (previously labelled variable cards seemingly no longer have a name):

```
[22May2021 12:44:43.251] [User Authenticator #4/INFO] [net.minecraft.network.login.ServerLoginNetHandler/]: UUID of player -snip-
[22May2021 12:45:05.965] [Server thread/INFO] [net.minecraft.server.management.PlayerList/]: -snip- logged in with entity id 143726 at (2668.733733639922, 67.0, -2384.892631026357)
[22May2021 12:45:05.976] [Server thread/INFO] [net.minecraft.server.dedicated.DedicatedServer/]: -snip- joined the game
[22May2021 12:45:06.005] [Server thread/INFO] [se.mickelus.tetra.data.DataManager/]: Sending data to client: -snip-
[22May2021 12:45:06.015] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: java.lang.NullPointerException
```

Doesn't look like the debug.log has any additional information about this, unfortunately. This is with version 1.8.2 of Integrated Dynamics, and we haven't really gotten very deep into the mod yet so there aren't many labelled variable cards.

**Comment by rubensworks** [comment]
@gjerm Could you send me the full log file? It looks like that stacktrace may explain the problem, but I need to see more lines below it.

**Comment by gjerm** [comment]
Apologies, the player join that I sent did not have any stacktrace under that line which is why I cut it off there. It looks like it sometimes gives a stacktrace and sometimes not. Here's a latest.log with both variations of the error:

[latest.log](https://github.com/CyclopsMC/IntegratedDynamics/files/6527837/latest.log)

**Comment by rubensworks** [comment]
Thanks!

Note to self:
```
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: java.lang.NullPointerException
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.network.PacketBuffer.func_211400_a(PacketBuffer.java:361)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.network.PacketBuffer.func_180714_a(PacketBuffer.java:357)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at org.cyclops.cyclopscore.network.PacketCodec$1.encode(PacketCodec.java:44)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at org.cyclops.cyclopscore.network.PacketCodec$9.encode(PacketCodec.java:178)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at org.cyclops.cyclopscore.network.PacketCodec.lambda$encode$0(PacketCodec.java:500)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at org.cyclops.cyclopscore.network.PacketCodec.loopCodecFields(PacketCodec.java:484)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at org.cyclops.cyclopscore.network.PacketCodec.encode(PacketCodec.java:493)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at org.cyclops.cyclopscore.network.PacketHandler.lambda$register$1(PacketHandler.java:61)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.fml.network.simple.IndexedMessageCodec.lambda$tryEncode$4(IndexedMessageCodec.java:134)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at java.util.Optional.ifPresent(Optional.java:159)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.fml.network.simple.IndexedMessageCodec.tryEncode(IndexedMessageCodec.java:132)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.fml.network.simple.IndexedMessageCodec.build(IndexedMessageCodec.java:147)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.fml.network.simple.SimpleChannel.encodeMessage(SimpleChannel.java:85)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.fml.network.simple.SimpleChannel.toBuffer(SimpleChannel.java:98)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.fml.network.simple.SimpleChannel.toVanillaPacket(SimpleChannel.java:129)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.fml.network.simple.SimpleChannel.sendTo(SimpleChannel.java:109)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at org.cyclops.cyclopscore.network.PacketHandler.sendToPlayer(PacketHandler.java:118)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at org.cyclops.integrateddynamics.core.persist.world.LabelsWorldStorage.onPlayerLoggedIn(LabelsWorldStorage.java:113)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.eventbus.ASMEventHandler_347_LabelsWorldStorage_onPlayerLoggedIn_PlayerLoggedInEvent.invoke(.dynamic)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.eventbus.ASMEventHandler.invoke(ASMEventHandler.java:85)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:302)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:283)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerLoggedIn(BasicEventHooks.java:44)
[23May2021 08:32:37.413] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.server.management.PlayerList.func_72355_a(PlayerList.java:231)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.network.login.ServerLoginNetHandler.func_147326_c(ServerLoginNetHandler.java:118)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.network.login.ServerLoginNetHandler.func_73660_a(ServerLoginNetHandler.java:65)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.network.NetworkManager.func_74428_b(NetworkManager.java:222)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.network.NetworkSystem.func_151269_c(NetworkSystem.java:134)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:865)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:291)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:787)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.server.MinecraftServer.func_240802_v_(MinecraftServer.java:642)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at net.minecraft.server.MinecraftServer.func_240783_a_(MinecraftServer.java:232)
[23May2021 08:32:37.414] [Server thread/INFO] [STDERR/]: [org.cyclops.cyclopscore.network.PacketCodec:loopCodecFields:487]: 	at java.lang.Thread.run(Thread.java:748)

```
